python/lsst/drp/tasks/wcsfit.py

```diff
@@ -0,0 +1,1019 @@
+    motion and parallax.
+
+    Parameters
+    ----------
+    refCat : `lsst.afw.table.SimpleCatalog`
+        Catalog including proper motion and parallax measurements
+    inputUnit : `astropy.unit.core.Unit`
+        Units of the input catalog
+    outputUnit : `astropy.unit.core.Unit`
+        Units required for the covariance matrix
+
+    Returns
+    -------
+    cov : `list` of `float`
+        Flattened output covariance matrix
+    """
+    # Here is the standard ordering of components in the cov matrix,
+    # to match the PM enumeration in C++ code of gbdes package's Match.
+    # Each tuple gives: the array holding the 1d error,
+    #                   the string in Gaia column names for this
+    #                   the ordering in the Gaia catalog
+    # and the ordering of the tuples is the order we want in our cov matrix
+    raErr = (refCat['coord_raErr'] * inputUnit).to(outputPMUnit).to_value()
+    decErr = (refCat['coord_decErr'] * inputUnit).to(outputPMUnit).to_value()
+    raPMErr = (refCat['pm_raErr'] * inputUnit).to(outputPMUnit).to_value()
+    decPMErr = (refCat['pm_decErr'] * inputUnit).to(outputPMUnit).to_value()
+    parallaxErr = (refCat['parallaxErr'] * inputUnit).to(outputPMUnit).to_value()
+    stdOrder = ((raErr, 'ra', 0),
+                (decErr, 'dec', 1),
+                (raPMErr, 'pmra', 3),
+                (decPMErr, 'pmdec', 4),
+                (parallaxErr, 'parallax', 2))
+    cov = np.zeros((len(refCat), 25))
+    k = 0
+    # TODO: when DM-35130, is done, we need the full covariance here
+    for i, pr1 in enumerate(stdOrder):
+        for j, pr2 in enumerate(stdOrder):
+            if pr1[2] < pr2[2]:
+                # add correlation coefficient (once it is available)
+                # cov[:, k] = (pr1[0] * pr2[0] * refCat[pr1[1] + '_' + pr2[1]
+                #              + '_corr'])
+                cov[:, k] = 0
+            elif pr1[2] > pr2[2]:
+                # add correlation coefficient (once it is available)
+                # cov[:, k] = (pr1[0] * pr2[0] * refCat[pr2[1] + '_' + pr1[1]
+                #              + '_corr'])
+                cov[:, k] = 0
+            else:
+                # diagnonal element
+                cov[:, k] = pr1[0] * pr2[0]
+            k = k+1
+
+    return cov
+
+
+def convertToAstPolyMapCoefficients(coefficients):
+    """Convert vector of polynomial coefficients from the format used in
+    `gbdes` into AST format (see Poly2d::vectorIndex(i, j) in
+    gbdes/gbutil/src/Poly2d.cpp). This assumes two input and two output
+    coordinates.
+
+    Parameters
+    ----------
+    coefficients : `list`
+        Coefficients of the polynomials
+    degree : `int`
+        Degree of the polynomial.
+
+    Returns
+    -------
+    astPoly : `astshim.PolyMap`
+        Coefficients in AST polynomial format
+    """
+    polyArray = np.zeros((len(coefficients), 4))
+    N = len(coefficients) / 2
+    degree = int(-1.5 + 0.5 * (1 + 8 * N)**0.5)
```

Comment:
What is this math? I think this definitely needs a comment explaining what is going on here; that int conversion of an inherently floating point number (sqrt of an odd number?) worries me.

Reply:
This is the quadratic formula applied to the formula for finding the number of coefficients of a multivariable polynomial up to a certain degree (I forget what the latter formula is called). It's guaranteed to be an integer. I added a comment.